substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/code/CodeInfoEncoder.java

```diff
@@ -393,6 +577,286 @@ public boolean verifyFrameInfo(CodeInfo info) {
+            buf.putU2(type.members().size());
+            for (Map.Entry<String, Method> entry : type.members().entrySet()) {
+                String memberName = entry.getKey();
+                Method valueAccessor = entry.getValue();
+                buf.putU2(encoders.sourceMethodNames.getIndex(memberName));
+                encodeValue(buf, valueAccessor.invoke(annotation), type.memberTypes().get(memberName));
+            }
+        }
+
+        void encodeValue(UnsafeArrayTypeWriter buf, Object value, Class<?> type) throws InvocationTargetException, IllegalAccessException {
+            buf.putU1(tag(type));
+            if (type.isAnnotation()) {
+                encodeAnnotation(buf, (Annotation) value);
+            } else if (type.isEnum()) {
+                buf.putU2(encoders.sourceClasses.getIndex(type));
+                buf.putU2(encoders.sourceMethodNames.getIndex(((Enum<?>) value).name()));
+            } else if (type.isArray()) {
+                encodeArray(buf, value, type.getComponentType());
+            } else if (type == Class.class) {
+                buf.putU2(encoders.sourceClasses.getIndex((Class<?>) value));
+            } else if (type == String.class) {
+                buf.putU2(encoders.sourceMethodNames.getIndex((String) value));
+            } else if (type.isPrimitive() || Wrapper.isWrapperType(type)) {
+                Wrapper wrapper = type.isPrimitive() ? Wrapper.forPrimitiveType(type) : Wrapper.forWrapperType(type);
+                switch (wrapper) {
+                    case BOOLEAN:
+                        buf.putU1((boolean) value ? 1 : 0);
+                        break;
+                    case BYTE:
+                        buf.putS1((byte) value);
+                        break;
+                    case SHORT:
+                        buf.putS2((short) value);
+                        break;
+                    case CHAR:
+                        buf.putU2((char) value);
+                        break;
+                    case INT:
+                        buf.putS4((int) value);
+                        break;
+                    case LONG:
+                        buf.putS8((long) value);
+                        break;
+                    case FLOAT:
+                        buf.putS4(Float.floatToRawIntBits((float) value));
+                        break;
+                    case DOUBLE:
+                        buf.putS8(Double.doubleToRawLongBits((double) value));
+                        break;
+                    default:
+                        throw shouldNotReachHere();
+                }
+            } else {
+                throw shouldNotReachHere();
+            }
+        }
+
+        void encodeArray(UnsafeArrayTypeWriter buf, Object value, Class<?> componentType) throws InvocationTargetException, IllegalAccessException {
+            if (!componentType.isPrimitive()) {
+                Object[] array = (Object[]) value;
+                buf.putU2(array.length);
```

Comment:
Who guarantees that these array lengths fit in an `U2`?

Reply:
The JDK decoder reads those as shorts so there must be an invariant in the encoding code requiring it.